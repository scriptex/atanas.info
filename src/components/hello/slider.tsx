import * as React from 'react';
import useInterval from 'use-interval';

import { titles } from '../../data/titles';

export const Slider: React.FC = () => {
	const [activeIndex, setActiveIndex] = React.useState(0);

	useInterval(() => {
		setActiveIndex(activeIndex === titles.length - 1 ? 0 : activeIndex + 1);
	}, 5000);

	return (
		<div className="c-slider">
			<h1>
				ðŸ‘‹ Hello! <br />I am Atanas Atanasov
			</h1>

			<h2>
				{titles.map((title: string, index: number) => (
					<span key={index} className={`${index === activeIndex ? 'current' : ''}`}>
						{title}
					</span>
				))}
			</h2>
		</div>
	);
};

export default Slider;
