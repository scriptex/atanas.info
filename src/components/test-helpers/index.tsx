import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '../../scripts/shared';

export const mockUseEffect = (): void => {
	let useEffect: jest.Mock<React.EffectCallback>;

	const mockedUseEffect = () => {
		useEffect.mockImplementationOnce((f: () => React.EffectCallback) => f());
	};

	beforeEach(() => {
		useEffect = jest.spyOn(React, 'useEffect') as unknown as jest.Mock<React.EffectCallback>;

		mockedUseEffect();
	});
};

export const snapshotTest = (
	Component: React.FC<Readonly<React.ComponentProps<any>>>,
	buttonSelector?: string
): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	mockUseEffect();

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { asFragment, container } = result;

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				act(() => {
					container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			let result!: RenderResult;

			await act(async () => {
				result = await waitFor(() => render(<Component />));
			});

			const { unmount, asFragment } = result;

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};
