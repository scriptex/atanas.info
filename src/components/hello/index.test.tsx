import { Hello } from '.';
import { Slider } from './slider';
import { snapshotTest } from '../test-helpers';

snapshotTest(Hello);
snapshotTest(Slider);
