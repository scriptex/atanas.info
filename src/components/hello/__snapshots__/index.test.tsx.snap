// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionHello component Should render the SectionHello component 1`] = `
<DocumentFragment>
  <main
    class="o-main o-main--high"
  >
    <section
      class="c-section c-section--hello"
      id="hello"
    >
      <div
        class="c-canvas"
        id="canvas"
      />
      <div
        class="c-slider"
      >
        <h1>
          ðŸ‘‹ Hello! 
          <br />
          I am Atanas Atanasov
        </h1>
        <h2>
          <span
            class="current"
          >
            lead front-end developer at E.ON
          </span>
          <span
            class=""
          >
            senior software engineer at Toptal
          </span>
          <span
            class=""
          >
            founding member at Braintrust
          </span>
          <span
            class=""
          >
            a TypeScript fanatic
          </span>
          <span
            class=""
          >
            a CSS wizard
          </span>
          <span
            class=""
          >
            a JavaScript tinkerer
          </span>
          <span
            class=""
          >
            a minimalist
          </span>
          <span
            class=""
          >
            an open source enthusiast
          </span>
          <span
            class=""
          >
            a maker of things
          </span>
          <span
            class=""
          >
            an a11y and UX advocate
          </span>
          <span
            class=""
          >
            a front-end magician
          </span>
          <span
            class=""
          >
            a bug fixer
          </span>
          <span
            class=""
          >
            a senior web developer
          </span>
          <span
            class=""
          >
            I am into data visualization
          </span>
          <span
            class=""
          >
            an UI architect
          </span>
          <span
            class=""
          >
            I am obsessed with quality
          </span>
          <span
            class=""
          >
            a web standards fanatic
          </span>
          <span
            class=""
          >
            a JavaScript master
          </span>
          <span
            class=""
          >
            a NodeJS hacker
          </span>
          <span
            class=""
          >
            a father
          </span>
          <span
            class=""
          >
            a husband
          </span>
        </h2>
      </div>
    </section>
  </main>
</DocumentFragment>
`;

exports[`SectionHello component Should render the SectionHello component 2`] = `<DocumentFragment />`;

exports[`Slider component Should render the Slider component 1`] = `
<DocumentFragment>
  <div
    class="c-slider"
  >
    <h1>
      ðŸ‘‹ Hello! 
      <br />
      I am Atanas Atanasov
    </h1>
    <h2>
      <span
        class="current"
      >
        lead front-end developer at E.ON
      </span>
      <span
        class=""
      >
        senior software engineer at Toptal
      </span>
      <span
        class=""
      >
        founding member at Braintrust
      </span>
      <span
        class=""
      >
        a TypeScript fanatic
      </span>
      <span
        class=""
      >
        a CSS wizard
      </span>
      <span
        class=""
      >
        a JavaScript tinkerer
      </span>
      <span
        class=""
      >
        a minimalist
      </span>
      <span
        class=""
      >
        an open source enthusiast
      </span>
      <span
        class=""
      >
        a maker of things
      </span>
      <span
        class=""
      >
        an a11y and UX advocate
      </span>
      <span
        class=""
      >
        a front-end magician
      </span>
      <span
        class=""
      >
        a bug fixer
      </span>
      <span
        class=""
      >
        a senior web developer
      </span>
      <span
        class=""
      >
        I am into data visualization
      </span>
      <span
        class=""
      >
        an UI architect
      </span>
      <span
        class=""
      >
        I am obsessed with quality
      </span>
      <span
        class=""
      >
        a web standards fanatic
      </span>
      <span
        class=""
      >
        a JavaScript master
      </span>
      <span
        class=""
      >
        a NodeJS hacker
      </span>
      <span
        class=""
      >
        a father
      </span>
      <span
        class=""
      >
        a husband
      </span>
    </h2>
  </div>
</DocumentFragment>
`;

exports[`Slider component Should render the Slider component 2`] = `<DocumentFragment />`;
